File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/dfs_rec.py
"""Recursive Depth First Search"""

from graphs import G1, init_visited


def dfs(graph: list[str, list[str]], start_node: str, visited: dict[str, bool]):
    """Recursive Depth First Search starting from the start node using a queue."""
    visited[start_node] = True
    print(start_node) # Process
    for neighbour in graph[start_node]:
        if not visited[neighbour]:
            dfs(graph, neighbour, visited)


def main():
    dfs(graph=G1, start_node="A", visited=init_visited(G1))


if __name__ == "__main__":
    main()

File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/dfs_rec.cpp
#include "graphs.h"

// initVisited maps all nodes as unvisited.
visitedMap(string) initVisited(adjacencyMap(string) graph) {
    visitedMap(string) visited;
    for (auto const&[node, neighbous] : graph) {
        visited[node] = false;
    }
    return visited;
}

// Recursive depth first search from the starting node using a queue.
void dfs(adjacencyMap(string) graph, string node, visitedMap(string) *visited) {
    (*visited)[node] = true;
    cout << node << "\n"; // Process Node.

    for (string neighbour: graph[node]){
        if (!(*visited)[neighbour]) {
            dfs(graph, neighbour, visited);
        }
    }
}

int main(void) {
    visitedMap(string) visited = initVisited(G1);
    dfs(G1, "A", &visited);
    return EXIT_SUCCESS;
}
